### PR TITLE
Output constraint

### PR DESCRIPTION
**Pull request by oluowoj**
Missing "physical constraint" column added to Output constraint table in gamephy SRS. Please merge changes.

image

Still need to address an old comment in the code, will do in new pull request, if branch gets deleted

image

### CODE REVIEW

**Comment by oluowoj** [comment]
> 0.502 ? 9.8 ?

0.505 is the y coordinate position, to be fixed, not sure what you are asking about 9.8.

**Comment by JacquesCarette** [comment]
Sorry for the over-brief question/comment: what I meant is, what are those numbers and where do they come from? That information should be in Drasil, somehow.

The gravitational constant, for example, really ought to be in `drasil-data`, given to many more digits, and with provenance information.

File: code/drasil-example/Drasil/GamePhysics/Unitals.hs [hunk 1]
```diff
@@ -286,12 +287,17 @@ mmntOfInCons   = constrained' QP.momentOfInertia    [nonNegativeConstraint] (dbl
...
 posCons        = constrained' QP.position           [] (dbl 0.412) --FIXME: should be (0.412, 0.502) vector
 veloCons       = constrained' QP.velocity           [] (dbl 2.51)
-orientCons     = constrained' QM.orientation        [] (sy QM.pi_ / 2) -- physical constraint not needed space is radians
+orientCons     = constrained' QM.orientation        [physc $ Bounded (Inc, 0) (Inc, 2 * sy QM.pi_)] (sy QM.pi_ / 2) -- physical constraint not needed space is radians
 angVeloCons    = constrained' QP.angularVelocity    [] (dbl 2.1)
 forceCons      = constrained' QP.force              [] (dbl 98.1)
 torqueCons     = constrained' QP.torque             [] (dbl 200)
 restCoefCons   = constrained' QP.restitutionCoef    [physc $ Bounded (Inc,0) (Inc,1)] (dbl 0.8)
 
+posOutCons        = constrained' QP.position           [] (dbl 0.0)
```
Comment (JacquesCarette) [inline code/drasil-example/Drasil/GamePhysics/Unitals.hs]: Also, it makes no sense to me to use `constrained'` with an empty list of constraints! That, in fact, should be a hard error.

File: code/drasil-example/Drasil/GamePhysics/Unitals.hs [hunk 2]
```diff
@@ -286,12 +287,17 @@ mmntOfInCons   = constrained' QP.momentOfInertia    [nonNegativeConstraint] (dbl
 gravAccelCons  = constrained' QP.gravitationalConst [] (dbl 9.8)
```
Comment (smiths) [inline code/drasil-example/Drasil/GamePhysics/Unitals.hs]: Newton's gravitational constant is NOT the same thing as the acceleration due to gravity.  The value should be 6.67430(15)×10^{−11} (in the units used).  You can see this at https://en.wikipedia.org/wiki/Gravitational_constant .  @JacquesCarette is correct that this value, and the acceleration due to gravity on the Earth (9.81 m^2/2) should be stored somewhere in Drasil that all of the examples can access.  I don't know how we add  provenance information (I've been lumping all of this kind of information into the term "rationale"), but we should definitely start thinking about this.  It keeps coming up, which is a sign to me that we are on the right track.  :smile:

**Comment by oluowoj** [comment]
@bmaclach do you know where we store these actual numbers, looks like there is some input and output constraint table. Could you point me to where these values are stored please?

**Comment by bmaclach** [comment]
@oluowoj I'm not sure I know which "values" you are referring to, but if you mean the constraints that show in the input/output constraint tables, those come from the `ConstrConcept` for the inputs and outputs, which are defined in Unitals.hs for your example.

**Comment by oluowoj** [comment]
> @oluowoj I'm not sure I know which "values" you are referring to, but if you mean the constraints that show in the input/output constraint tables, those come from the ConstrConcept for the inputs and outputs, which are defined in Unitals.hs for your example.

I see, I will take a look at this. thx

**Comment by oluowoj** [comment]
> @oluowoj I'm not sure I know which "values" you are referring to, but if you mean the constraints that show in the input/output constraint tables, those come from the ConstrConcept for the inputs and outputs, which are defined in Unitals.hs for your example.


@Brooks, I was actually referring to constant values of some of the constraint fields. Remember how we were talking about constants last week in the All hands meeting and I mentioned that the value of pi should be somewhere in Drasil(I think I've seen it but not too sure).
 
So yes, constraint values for game physics are currently in Unitals.hs, however, some of those values are actually constants, which I suppose might have been stored as well and I need not rewrite in Unitals but just reuse from wherever its being stored....and maybe it currently does not exist.

**Comment by bmaclach** [comment]
@oluowoj Okay, thanks for the clarification. The "quantity" for pi is stored in drasil-data/Data/Drasil/Quantities/Math.hs. Looks like the gravitational acceleration unital is defined in drasil-data/Data/Drasil/Quantities/Physics.hs. That's probably what you want. But note that the actual value is not defined anywhere. So if you use that unital, it will be the symbol `g` that shows up in your constraint, not the value `9.8`. Is that okay?

**Comment by oluowoj** [comment]
> @oluowoj Okay, thanks for the clarification. The "quantity" for pi is stored in drasil-data/Data/Drasil/Quantities/Math.hs. Looks like the gravitational acceleration unital is defined in drasil-data/Data/Drasil/Quantities/Physics.hs. That's probably what you want. But note that the actual value is not defined anywhere. So if you use that unital, it will be the symbol g that shows up in your constraint, not the value 9.8. Is that okay?

oh okay, I need the actual value which I think we should consider storing somewhere in Drasil since these are values that can be reused in other examples. @JacquesCarette mentioned above that gravitational constant be somewhere in drasil-data but maybe he wasn't referring to the values. @JacquesCarette, did you mean the actual values like 9.8 and so on?

**Comment by oluowoj** [comment]
> @oluowoj Okay, thanks for the clarification. The "quantity" for pi is stored in drasil-data/Data/Drasil/Quantities/Math.hs. Looks like the gravitational acceleration unital is defined in drasil-data/Data/Drasil/Quantities/Physics.hs. That's probably what you want. But note that the actual value is not defined anywhere. So if you use that unital, it will be the symbol g that shows up in your constraint, not the value 9.8. Is that okay?
> 
> oh okay, I need the actual value which I think we should consider storing somewhere in Drasil since these are values that can be reused in other examples. @JacquesCarette mentioned above that gravitational constant should be somewhere in drasil-data but maybe he wasn't referring to the values. @JacquesCarette, did you mean the actual values of constants like 9.8 and so on?

**Comment by JacquesCarette** [comment]
Yes, we should have these values in Drasil. Weirdly, I don't think we have the notion of a 'constant'!  [Yes, I meant that value].  What have Quantities with reasonable values, etc, but no 'constant'. Should be added.

**Comment by oluowoj** [comment]
please hold on this PR, it is dependent on   issue #1592

**Comment by JacquesCarette** [comment]
This is still failing Travis for log differences. And of course, since @danscime has made changes here too (renamings), now things are out of sync again...

**Comment by oluowoj** [comment]
Trying to display the value of G as 
image
But I get this :
image
Looks like when the exponent digit is more than 1, it translates it differently. 
Here is the code: 
image

Any clues?

**Comment by bmaclach** [comment]
@oluowoj That is expected behaviour as we use engineering notation as described here: https://github.com/JacquesCarette/Drasil/pull/1496#discussion_r291226461

**Comment by oluowoj** [comment]
> These kinds of corrective fixes really should have been done as a new PR, not on top of the old one. Many small PRs are way better than a single evolving but never quite finishing monster of a PR.

This has been moved to PR #1746

**Comment by JacquesCarette** [comment]
This still doesn't build, even after the latest merge.  Please try to stay on top of these things. [And yet another merge is needed too]

**Comment by oluowoj** [comment]
> This still doesn't build, even after the latest merge.  Please try to stay on top of these things. [And yet another merge is needed too]

I am on top of it, but still figuring out why the build keeps failing.

**Comment by oluowoj** [comment]
> Indeed - but no other changes were made to this file. Was it not compiling before?

 I don't think its something new that I just added, it was there before and without this import the file does not compile.

**Pull request merged**